### PR TITLE
Add hook to new close_range function all, solves #439

### PR DESCRIPTION
**Pull request by recolic**

### CODE REVIEW

**Comment by recolic** [comment]
Tested and confirmed that `proxychains ssh xxx.com` is working

**Comment by recolic** [comment]
There is a bug.. I will reopen this PR after fixing this bug.

**Comment by rofl0r** [comment]
what happens if the user has a libc version without close_range ?

**Comment by recolic** [comment]
> what happens if the user has a libc version without close_range ?

I don't know... Let me test it.

File: src/libproxychains.c [hunk 1]
```diff
@@ -587,6 +600,51 @@ static int is_v4inv6(const struct in6_addr *a) {
...
+        if(close_range_buffer_cnt >= (sizeof close_range_buffer / sizeof *close_range_buffer)) {
+            errno = ENOMEM;
+            return -1;
+        }
+        int i = close_range_buffer_cnt++;
+        close_range_buffer[i].first = first;
+        close_range_buffer[i].last = last;
+        close_range_buffer[i].flags = flags;
+        return errno = 0;
+    }
```
Comment (rofl0r) [inline src/libproxychains.c:621]: there are various whitespace bugs like this one here using spaces instead of tabs

File: src/libproxychains.c [hunk 2]
```diff
@@ -587,6 +600,51 @@ static int is_v4inv6(const struct in6_addr *a) {
...
+	if(proxychains_resolver != DNSLF_RDNS_THREAD) return true_close_range(first, last, flags);
+
+	/* prevent rude programs (like ssh) from closing our pipes */
+    int res = 0, uerrno = 0;
+    int protected_fds[] = {req_pipefd[0], req_pipefd[1], resp_pipefd[0], resp_pipefd[1]};
+    int protected_fds_size = sizeof protected_fds / sizeof *protected_fds;
```
Comment (rofl0r) [inline src/libproxychains.c:627]: i'm not sure `sizeof *protected` is right here as it's an array, not a pointer.
should be `sizeof xxx[0]`

File: src/libproxychains.c [hunk 3]
```diff
@@ -587,6 +600,51 @@ static int is_v4inv6(const struct in6_addr *a) {
...
+    qsort(protected_fds, protected_fds_size, sizeof *protected_fds, compare_func_int);
+    /* We are skipping protected_fds while calling true_close_range() */
+    for(int i = 0; i < protected_fds_size; ++i) {
+        if(protected_fds[i] < first || protected_fds[i] > last)
+            continue;
+        int prev = (i == 0 || protected_fds[i-1] < first) ? first : protected_fds[i-1]+1;
+        int next = (i == protected_fds_size-1 || protected_fds[i+1] > last) ? last : protected_fds[i+1]-1;
+        if(prev != protected_fds[i]) {
+		    res |= true_close_range(prev, protected_fds[i]-1, flags);
+            uerrno = errno == 0 ? uerrno : errno;
```
Comment (rofl0r) [inline src/libproxychains.c:637]: you should only access errno iff the call to true_close_range returned error in that specific invocation

File: src/libproxychains.c [hunk 4]
```diff
@@ -858,5 +917,6 @@ DYLD_HOOK(freeaddrinfo);
 DYLD_HOOK(gethostbyaddr);
 DYLD_HOOK(getnameinfo);
 DYLD_HOOK(close);
+DYLD_HOOK(close_range);
```
Comment (rofl0r) [inline src/libproxychains.c:920]: monterey hooking is only for OSX and afaik it doesn't have close_range

File: src/libproxychains.c (hunk 1 above)
Comment (recolic) [reply src/libproxychains.c:621]: TODO: will fix this after fixing other issues

File: src/libproxychains.c (hunk 2 above)
Comment (recolic) [reply src/libproxychains.c:627]: Copied from this example: https://en.cppreference.com/w/c/algorithm/qsort
I believe it's ok.

File: src/libproxychains.c (hunk 4 above)
Comment (recolic) [reply src/libproxychains.c:920]: Thanks for this knowledge! fixed in latest commit.

File: src/libproxychains.c [hunk 5]
```diff
@@ -587,6 +600,51 @@ static int is_v4inv6(const struct in6_addr *a) {
...
+    }
+	if(proxychains_resolver != DNSLF_RDNS_THREAD) return true_close_range(first, last, flags);
+
+	/* prevent rude programs (like ssh) from closing our pipes */
+    int res = 0, uerrno = 0;
+    int protected_fds[] = {req_pipefd[0], req_pipefd[1], resp_pipefd[0], resp_pipefd[1]};
+    int protected_fds_size = sizeof protected_fds / sizeof *protected_fds;
+    qsort(protected_fds, protected_fds_size, sizeof *protected_fds, compare_func_int);
+    /* We are skipping protected_fds while calling true_close_range() */
+    for(int i = 0; i < protected_fds_size; ++i) {
```
[the remark this answers is not in the document]
Comment (recolic) [reply src/libproxychains.c:630]: Fixed. Correctness proof: https://paste.ee/p/39ySs

**Comment by recolic** [comment]
Algorithm correctness problem fixed. Tested at https://paste.ee/p/39ySs

File: src/libproxychains.c (hunk 3 above)
Comment (recolic) [reply src/libproxychains.c]: fixed in latest commit

**Comment by recolic** [comment]
I added a `is_mandatory` parameter to load_sym, to make the close_range optional. 

I have tested and confirm that, it's working on both ubuntu 16.04 and latest archlinux. 

I'm not sure if it's the best solution. Could you review it? Thanks

File: src/libproxychains.c [hunk 6]
```diff
@@ -587,6 +600,64 @@ static int is_v4inv6(const struct in6_addr *a) {
...
+}
+/* Warning: Linux manual says the third arg is `unsigned int`, but unistd.h says `int`. */
+HOOKFUNC(int, close_range, unsigned first, unsigned last, int flags) {
+	if(true_close_range == NULL) {
+	    fprintf(stderr, "Calling close_range, but this platform does not provide this system call. ");
+	    return -1;
+	}
+	if(!init_l) {
+	    /* push back to cache, and delay the execution. */
+	    if(close_range_buffer_cnt >= (sizeof close_range_buffer / sizeof *close_range_buffer)) {
```
Comment (rofl0r) [inline src/libproxychains.c:616]: this still uses `*xxx` instead of `xxx[0]`

File: src/libproxychains.c [hunk 7]
```diff
@@ -587,6 +600,64 @@ static int is_v4inv6(const struct in6_addr *a) {
...
+	int protected_fds[] = {req_pipefd[0], req_pipefd[1], resp_pipefd[0], resp_pipefd[1]};
+	int protected_fds_size = sizeof protected_fds / sizeof protected_fds[0];
+	qsort(protected_fds, protected_fds_size, sizeof protected_fds[0], compare_func_int);
+	/* We are skipping protected_fds while calling true_close_range()
+	 * If protected_fds cut the range into some sub-ranges, we close sub-ranges BEFORE cut point in the loop. 
+	 * [first, cut1-1] , [cut1+1, cut2-1] , [cut2+1, cut3-1]
+	 * Finally, we delete the remaining sub-range, outside the loop. [cut3+1, tail]
+	 */
+	int next_fd_to_close = first;
+	for(int i = 0; i < protected_fds_size; ++i) {
```
Comment (rofl0r) [inline src/libproxychains.c:639]: pls move the int i declaration outside the loop, we dont use that feature anywhere else in code

File: src/libproxychains.c [hunk 8]
```diff
@@ -587,6 +600,64 @@ static int is_v4inv6(const struct in6_addr *a) {
...
+	    if(protected_fds[i] < first || protected_fds[i] > last)
+	        continue;
+	    int prev = (i == 0 || protected_fds[i-1] < first) ? first : protected_fds[i-1]+1;
+	    if(prev != protected_fds[i]) {
+	        if(-1 == true_close_range(prev, protected_fds[i]-1, flags)) {
+	            res = -1;
```
Comment (rofl0r) [inline src/libproxychains.c:645]: this still uses mixed indentation, codestyle is tab only

File: src/libproxychains.c [hunk 9]
```diff
@@ -631,7 +702,7 @@ HOOKFUNC(int, connect, int sock, const struct sockaddr *addr, unsigned int len)
 	PDEBUG("port: %d\n", port);
 
 	// check if connect called from proxydns
-        remote_dns_connect = !v6 && (ntohl(p_addr_in->s_addr) >> 24 == remote_dns_subnet);
+	    remote_dns_connect = !v6 && (ntohl(p_addr_in->s_addr) >> 24 == remote_dns_subnet);
```
Comment (rofl0r) [inline src/libproxychains.c:705]: please dont change indentation in unrelated code

File: src/libproxychains.c [hunk 10]
```diff
@@ -290,7 +303,7 @@ static void get_chain_data(proxy_data * pd, unsigned int *proxy_count, chain_typ
 	if( ( file = fopen(env, "r") ) == NULL )
 	{
 	        perror("couldnt read configuration file");
-        	exit(1);
+	    	exit(1);
```
Comment (rofl0r) [inline src/libproxychains.c:306]: pls dont change indentation in unrelated code

File: src/libproxychains.c [hunk 11]
```diff
@@ -86,13 +87,13 @@ static int init_l = 0;
...
 
-static void* load_sym(char* symname, void* proxyfunc) {
+static void* load_sym(char* symname, void* proxyfunc, int is_mandatory) {
 
 	void *funcptr = dlsym(RTLD_NEXT, symname);
 
 	if(!funcptr) {
-		fprintf(stderr, "Cannot load symbol '%s' %s\n", symname, dlerror());
-		exit(1);
+		proxychains_write_log(LOG_PREFIX "Cannot load symbol '%s', is_mandatory=%d, %s\n", symname, is_mandatory, dlerror());
```
Comment (rofl0r) [inline src/libproxychains.c:95]: it's not necessary to change log function, iirc at this point fprintf is used because proxychains isnt fully set up yet.

Comment (rofl0r) [reply src/libproxychains.c:95]: i think if not mandatory we best dont log this at all. otherwise users without close_range will get spam whenever they use proxychains.

File: src/libproxychains.c (hunk 7 above)
Comment (recolic) [reply src/libproxychains.c]: fixed in latest commit.

File: src/libproxychains.c (hunk 10 above)
Comment (recolic) [reply src/libproxychains.c]: fixed in latest commit.

File: src/libproxychains.c (hunk 11 above)
Comment (recolic) [reply src/libproxychains.c]: fixed in latest commit. 
I have reverted to the old `fprintf`, and the error only shows for mandatory symbol.

File: src/libproxychains.c [hunk 12]
```diff
@@ -86,15 +87,14 @@ static int init_l = 0;
...
-
+static void* load_sym(char* symname, void* proxyfunc, int is_mandatory) {
 	void *funcptr = dlsym(RTLD_NEXT, symname);
 
-	if(!funcptr) {
-		fprintf(stderr, "Cannot load symbol '%s' %s\n", symname, dlerror());
+	if(is_mandatory && !funcptr) {
+		fprintf(stderr, "Cannot load symbol '%s', is_mandatory=%d, %s\n", symname, is_mandatory, dlerror());
 		exit(1);
 	} else {
```
Comment (rofl0r) [inline src/libproxychains.c:96]: `} else if (!funcptr) {return 0;`

File: src/libproxychains.c [hunk 13]
```diff
@@ -86,15 +87,14 @@ static int init_l = 0;
...
 
-static void* load_sym(char* symname, void* proxyfunc) {
-
+static void* load_sym(char* symname, void* proxyfunc, int is_mandatory) {
 	void *funcptr = dlsym(RTLD_NEXT, symname);
 
-	if(!funcptr) {
-		fprintf(stderr, "Cannot load symbol '%s' %s\n", symname, dlerror());
+	if(is_mandatory && !funcptr) {
+		fprintf(stderr, "Cannot load symbol '%s', is_mandatory=%d, %s\n", symname, is_mandatory, dlerror());
```
Comment (rofl0r) [inline src/libproxychains.c:94]: we don't need the changes here either, it will only be printed for mandatory funcs anyway

File: src/libproxychains.c (hunk 12 above)
Comment (recolic) [reply src/libproxychains.c:96]: if funcptr is NULL, it will return NULL at line 103. 

Is it necessary to explicitly write this line?

File: src/libproxychains.c [hunk 14]
```diff
@@ -86,15 +87,14 @@ static int init_l = 0;
...
 		exit(1);
 	} else {
-		PDEBUG("loaded symbol '%s'" " real addr %p  wrapped addr %p\n", symname, funcptr, proxyfunc);
+		PDEBUG("loaded symbol '%s' real addr %p wrapped addr %p\n", symname, funcptr, proxyfunc);
```
[the remark this answers is not in the document]
Comment (recolic) [reply src/libproxychains.c]: reverted the format enhancement in latest commit.

File: src/libproxychains.c (hunk 12 above)
Comment (rofl0r) [reply src/libproxychains.c:96]: to prevent the PDEBUG line from being printed

Comment (recolic) [reply src/libproxychains.c:96]: got it. I will do that

File: src/libproxychains.c (hunk 13 above)
Comment (recolic) [reply src/libproxychains.c]: fixed in latest commit.

File: src/libproxychains.c (hunk 12 above)
Comment (recolic) [reply src/libproxychains.c]: fixed in latest commit.

**Comment by rofl0r** [comment]
heh, how did you manage to produce a hash of 00000000 not only once, but twice ?

**Comment by recolic** [comment]
> heh, how did you manage to produce a hash of 00000000 not only once, but twice ?

That's interesting! Try my project: https://github.com/recolic/git-commit-hash-vanitygen
